### ISSUE TITLE
Live stream entity's missing from HA 2022.4.0

### ISSUE THREAD

**Issue by danielsmith89** [tsightler/ring-mqtt#349]
### Describe the bug
Since updating to Home Assistant 2022.4.0, the live stream entities for my Ring camera's/doorbell's have gone missing.

### Describe your environment
Home Assistant running on Docker

**Comment by tkernaghan** [comment]
Wish I read this before doing the upgrade, same problem here running on Home Assistant OS. I tried regenerating the token, which was successful, and rebooting the server a couple times but my live streams are still missing

**Comment by danielsmith89** [comment]
All my other entities are fine, it's just the live camera streams that I don't have. Hopefully, it should be a relatively simple fix @tsightler ?

**Comment by broyuken** [comment]
Same issue here. Live camera entities are gone 

2022-04-06 19:52:14 ERROR (MainThread) [homeassistant.components.generic.config_flow] Error importing generic IP camera platform config: unexpected error '['unable_still_load']'

**Comment by tsightler** [comment]
Again, this issue cannot be caused by any code in this project.  It has to be caused by a change in Home Assistant.  The release notes cover the changes to generic IP camera configuration and clearly from the logs above the import is failing.  A Google search immediately returns results from [other users experiencing the same issue](https://github.com/home-assistant/core/issues/69461) with other Generic IP cameras.  It seems like the issue is related to the inability to import the still_image_url so you could remove that from the config and restart and see if it will import, or you can try to set the camera up manually with the UI configuration, although it appears that this may no longer support using a template value, so I'm not sure if it will be possible to use the snapshot image anymore.

Regardless, this is all 100% HA configuration related, not anything to do with this project directly, so I'm closing this.

**Comment by danielsmith89** [comment]
For those who would like a useful answer:

I managed to fix this by adding through the UI and selecting TCP as the transport (which I suspect might be the problem with it trying to import the Yaml config as this isn't defined), the templated still_image_url worked too.

**Comment by broyuken** [comment]
Could you clarify a little bit? I'm trying to add it via the UI and have TCP selected but I'm getting an authorization error. 

![image](https://user-images.githubusercontent.com/1593391/162210279-559ca4c3-f225-4bc7-969d-e2d51d4af404.png)

**Comment by broyuken** [comment]
Also, this may be helpful, I see the following error in my ring-mqtt logs

```
2022-04-07T14:02:55.715Z ring-rtsp INF [RTSP] [conn 172.30.32.1:45314] opened
2022-04-07T14:02:55.716Z ring-rtsp INF [RTSP] [conn 172.30.32.1:45314] closed (invalid path name: can contain only alphanumeric characters, underscore, dot, tilde, minus or slash (780473d4e513_live:8554/a810872def16_live))
```

It looks like there may be ( ) added around the url now for some reason?

**Comment by tsightler** [comment]
> For those who would like a useful answer:

I'm not sure how directing people to the only resource that could possibly have the information to solve the problem is not useful.  Home Assistant issues are supported by the Home Assistant developers and community.  The only other option is to wait until I have time to personally upgrade my instances, read the release notes, work through the upgrade issues, Google the problems, find the bugs, etc. which probably will not be until sometime next week at the earliest.

If this had simply been brought up on the discussion groups, well, that's one thing, but this is the "issues" section of this project and you opened this as a bug, which is basically that same as asking me, as the maintainer of this project, to help with an issue that is not in my control.  This issue is not a bug in this project, Home Assistant changed how it configured generic IP cameras and broke some stuff, so the issue is there and can only be solved by using the resources there.

**Comment by tsightler** [comment]
Finally had the time to upgrade my instances and see this issue first hand.  Somehow I only just noticed that my very first comment, which was to try removing the still_image_url field from the configuration and re-start, never posted at all.

Hopefully, the changes in 2022.4.1 will make the import process more reliable, however, even with that it took my production instances about 5-6 reboots before all 14 of my camera instances imported successfully, but eventually they did.

@broyuken Hopefully you got your issue resolved already, but, as far as I can tell, the "username/password" doesn't work but you can still include the username/password in the RTSP URL as previously and it should work fine. 

I'll try to update the docs at some point with the new camera setup method but, personally, I really hate this UI based configuration, and I especially despise being forced into it.  It might be fine and easier for people with a handful of cameras, but when you have a lot of cameras, it's a total pain and, worse, it's much more difficult to perform bulk changes across a large number of devices, or to copy the config from once instance to another.

Regardless, I have verified that nothing is really broken with ring-mqtt or the addon with HA 2022.4.x, it's just the HA config that needs to be changed.  Hopefully people will work their way through it.

**Comment by broyuken** [comment]
Yep I can confirm that it works after putting the auth back in the rtsp url

**Comment by tkernaghan** [comment]
Can someone please explain this new migration process in a little more detail? I tried upgrading a couple weeks ago from 2022.3.8 to 2022.4.1 and lost all my camera feeds. So I reverted back to 2022.3.8 and waited to see if the problem would be fixed. Today I upgraded to 2022.4.7 and the same issue exists. 

I tried to manually add the generic camera details through the UI but I keep getting different errors. I'm using the exact same still image and live stream URLs that are in my configurations.yaml file and selecting TCP and entering the username and password. I either get an unknown error or a template error, neither of which I understand. Do I need to remove the generic entries from YAML before trying to add them through the UI?

I just restored everything back to 2022.3.8 as I have screens throughout my house that show all the camera live feeds and is the main reason I use Home Assistant.

I'm an IT professional but not an HA professional so please explain the steps in layman's terms...

Thanks!

**Comment by tsightler** [comment]
@tkernaghan There's really not much to it so I'm not sure how to help, but I'll give it a try.  Basically, there were previously two entries in configuration.yaml, now you have to put those same two entries in the HA UI.  No other changes should be required, however, the UI does do some fairly strict validation of the entered values while the configuration.yaml config was more forgiving (which is interesting because it feels like it's usually the other way around).  If the entries in configuration.yaml are correct, it should, theoretically, automatically create the camera entities in the UI.  I had marginal success with this, but after multiple reboots they all did finally come in.  If they fail, you can look at the logs under Configuration...Settings...Log and you should see messages for the reason.

Of course, it's not required to allow the automatic migration, you should be able to manually add the camera and just use the same two working values.  As you say that's not working for you either then I would suggest to try with only the Stream Source URL, leaving the Still Image URL blank, and see if that works as problems with the template URL are the most common because the template syntax is quite picky.  If the camera at least works without the Still Image URL then you can come back and figure out what is wrong with the Still Image URL.  Maybe a screenshot would help?

![image](https://user-images.githubusercontent.com/17784731/165584783-64c161fe-8567-427e-9580-08035d109e08.png)

That's really all there is to it, two entries, one if you don't use the Stream Source URL.  In my case I use an issued SSL certificate via Let's Encrypt so the still image URL is using https:// and I'm able to leave the Verify SSL Certificate enabled, but if you use a self-signed certificate, or the certificate name doesn't match the hostname, you'd need to uncheck that.

If you still have issues maybe post the previously working config?

**Comment by broyuken** [comment]
I'll add that I also had to have tcp selected as the transport for it to work, and make sure the credentials are in the url string, and not in the username/password boxes.

**Comment by tsightler** [comment]
If setting TCP worked for you that's a really good suggestion, but I don't know why that would be required as HA has supported automatic protocol detection for RTSP for quite a while.  The option to specify the protocol was always available with the yaml based config as well, but I never had to explicitly define it and that didn't change when I migrated to the UI based configuration.

I'm thinking I may enable both UDP and TCP protocols on the RTSP server for future versions though.  The only reason I didn't initially is because I saw significantly more artifacts with UDP during initial testing, but I think that might be better with recent versions of rtsp-simple-server.

**Comment by tkernaghan** [comment]
Do you need to delete the generic camera entries from configuration.yaml before trying to add them through the UI?

**Comment by tsightler** [comment]
It's not required, although if you leave them in the configuration.yaml then Home Assistant will keep trying to import them every time it is restarted, but, even if the automatic import/conversion worked, you'd still have to manually remove them from configuration.yaml.

If you are still having an issue that best way to help would be in posting more details about the configuration, what values you are using, what errors you are seeing.  Also, note that it is pretty much absolutely required to have the addon working when you add the camera because HA attempts to connect to the camera to validate the entered settings.

**Comment by tsightler** [comment]
@tkernaghan Another user posted in the [discussion group](https://github.com/tsightler/ring-mqtt/discussions/356) about a very similar issue where they were unable to use the still image URL.  This appears to be caused by another HA bug introduced as part of the UI based camera configuration, which I didn't see because my instance is using standard 443 port for SSL.  I think this is likely causing the "generic error" that is seen, especially if you are using a non-standard web port like the default 8123

I think I've figured out a workaround to stop the HA parser from doing silly things.  As an example, if your previous working URL was the following

`http://192.168.50.10:8123{{ states.camera.front_doorbell_snapshot.attributes.entity_picture }}`
You can modify it instead to
`http://192.168.50.10:8123/.{{ states.camera.front_doorbell_snapshot.attributes.entity_picture }}`

Adding the extra "/." to the path prior to the template seems to be enough to tickle the parser into recognizing the port properly but doesn't impact the actual computed path.

If that works for you as well then I'll try to update the documentation and make sure this makes it to the next update which is nearing release (I hope).

**Comment by tkernaghan** [comment]
Ok I think you might be on to something here because I am not using SSL and using port 8123. I'll test this out when I've got some time over the weekend. I need to make sure both the still image and live feed work, all my dashboards have a view of the still image that refreshes every 10 seconds then I can click on the picture glance card to get the live view of the cameras.

**Comment by tsightler** [comment]
Just to follow up on this, thanks to user @nebhead opening an issue with the HA core team this issue should be addressed in a future version of HA.  I'm not sure exactly which version, but the PR was created a merged in the last couple of days but I'm not sure if it was in time to make the 2022.05.x releases or not.  Hopefully the generic camera support in HA will stabilize over the next few weeks and these HA related issues will be behind us.

In the interim, adding the extra "/." to the path immediately after the port number should effectively work around the issue.

**Comment by tsightler** [comment]
Just verified that the patch is included in the Home Assistant 2022.05.0b5 beta release so that means it should make the 2022.05.0 release when it comes out of beta in the coming days.  Testing on that version indicates that the templates should work again with no required changes.

**Comment by tkernaghan** [comment]
I just did the upgrade from 2022.3.8 to 2022.5.1 and it went perfectly smooth, it automatically imported all of my Ring cameras (on the first boot) and everything is working exactly like it did before. I didn't have to manually add anything and all my dashboards are working. I should also note that my stream URLs are not using SSL, so that problem appears to be solved as well. Thanks for all the hard work everyone!

**Comment by tsightler** [comment]
I don't really know how I can help with this as the live stream entities
are not automatically created, they have to be manually configured in Home
Assistant and always have been so if they are missing this has to be a Home
Assistant issue.
